File: benchmark/src/jmh/java/com/dawidpawliczek/benchmark/InMemoryCommandLog.java
package com.dawidpawliczek.benchmark;

import com.dawidpawliczek.engine.ports.CommandLog;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public final class InMemoryCommandLog implements CommandLog {

    private final List<byte[]> entries = new ArrayList<>();

    @Override
    public synchronized void append(byte[] payload) {
        entries.add(payload.clone());
    }

    @Override
    public void sync() {}

    @Override
    public synchronized void replay(Consumer<byte[]> handler) {
        for (byte[] entry : entries) {
            handler.accept(entry.clone());
        }
    }
}

File: app/src/main/kotlin/adapter/InMemoryCommandLog.kt
package com.dawidpawliczek.app.adapter

import com.dawidpawliczek.engine.ports.CommandLog
import java.util.function.Consumer

class InMemoryCommandLog : CommandLog {
    private val entries = ArrayList<ByteArray>()

    @Synchronized
    override fun append(payload: ByteArray) {
        entries.add(payload.copyOf())
    }

    override fun sync() { }

    @Synchronized
    override fun replay(handler: Consumer<ByteArray>) {
        entries.forEach { handler.accept(it.copyOf()) }
    }
}
